File: Python/515 Find Largest Value in Each Tree Row.py
# Find Largest Value in Each Tree Row
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        queue = [root]
        queue2 = []
        ans = [root.val]
        while queue:
            curr = queue.pop(0)
            if curr.left:
                queue2.append(curr.left)           
            if curr.right:
                queue2.append(curr.right)
            if not queue and queue2:
                ans.append(max(ele.val for ele in queue2))
                while queue2:
                    queue.append(queue2.pop(0))
        return ans

File: C++/515 Find Largest Value in Each Tree Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    vector<int> ans;
public:
    vector<int> largestValues(TreeNode* root) {
        return dfs(root, 0);
    }

    void dfs(TreeNode* node, int level) {
        if (node == nullptr) return;
        if (level ==  ans.size()) ans.push_back(node->val);
        else {
            ans[level] = max(ans[level], node->val);
        }
        dfs(node->left, level+1);
        dfs(node->right, level+1);
    }
    // vector<int> largestValues(TreeNode* root) {
    //     vector<int> ans;
    //     if (!root) return ans;
    //     queue<TreeNode*> q;
    //     q.push(root);
    //     while (!q.empty()) {
    //         queue<TreeNode*> level;
    //         int tmp = INT_MIN;
    //         while (!q.empty()) {
    //             TreeNode* curr = q.front();
    //             q.pop();
    //             tmp = max(tmp, curr->val);
    //             if (curr->left) level.push(curr->left);
    //             if (curr->right) level.push(curr->right);
    //         }
    //         ans.push_back(tmp);
    //         q = level;
    //     }
    //     return ans;
    // }
};